src/embit/bip47.py

```diff
@@ -0,0 +1,258 @@
+import hashlib
+import hmac
+import sys
+
+from binascii import hexlify, unhexlify
+from io import BytesIO
+from typing import Tuple
+
+from . import base58, ec, script
+from .base import EmbitError
+from .bip32 import HDKey
+from .networks import NETWORKS
+from .script import OPCODES
+from .transaction import Transaction
+if sys.implementation.name == "micropython":
+    import secp256k1
+else:
+    from .util import secp256k1
+
+
+"""
+    BIP-47: https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki
+"""
+
+class BIP47Exception(Exception):
+    pass
+
+
+def get_payment_code(root: HDKey, coin: int = 0, account: int = 0) -> str:
+    """
+        Generates the recipient's BIP-47 shareable payment code (version 1)
+        for the input root private key.
+    """
+    bip47_child = root.derive(f"m/47'/{coin}'/{account}'")
+
+    buf = BytesIO()
+    buf.write(b'\x01')      # bip47 version
+    buf.write(b'\x00')      # Bitmessage; always zero
+    buf.write(bip47_child.get_public_key().serialize())
+    buf.write(bip47_child.chain_code)
+    buf.write(b'\00' * 13)  # bytes reserved for future expansion
+
+    return base58.encode_check(b'\x47' + buf.getvalue())
+
+
+def get_derived_payment_code_node(payment_code: str, derivation_index: int, version: str = NETWORKS["main"]["xpub"]) -> HDKey:
+    """Returns the nth derived child for the payment_code"""
+    raw_payment_code = base58.decode_check(payment_code)
+
+    # 81-byte payment code format:
+    #   0x47 0x01 0x00 (sign) (32-byte pubkey) (32-byte chain code) (13 0x00 bytes)
+    pubkey = ec.PublicKey.from_string(hexlify(raw_payment_code[3:36]))
+    chain_code = raw_payment_code[36:68]
+    root = HDKey(key=pubkey, chain_code=chain_code, version=version)
+    return root.derive([derivation_index])
+
+
+def get_notification_address(payment_code: str, script_type: str = "p2pkh", network: str = NETWORKS["main"]) -> str:
+    """Returns the BIP-47 notification address associated with the given payment_code"""
+    # Get the 0th public key derived from the payment_code
+    pubkey = get_derived_payment_code_node(payment_code, derivation_index=0, version=network["xpub"]).get_public_key()
+
+    # TODO: Should we limit to just p2pkh?
+    if script_type == "p2pkh":
+        return script.p2pkh(pubkey).address(network)
+    elif script_type == "p2wpkh":
+        return script.p2wpkh(pubkey).address(network)
+    else:
+        raise EmbitError(f"Unsupported script_type: {script_type}")
+
+
+def get_payment_address(payer_root: HDKey, recipient_payment_code: str, index: int, coin: int = 0, account: int = 0, network=NETWORKS["main"], script_type: str = "p2wpkh") -> str:
```

Comment:
`script_type` should be `p2pkh`, otherwise, your test doesn't pass.

Reply:
Ah, I've been making bigger tweaks to the test suite. Fixed in incoming commit.